### ISSUE TITLE
Need an obvious way to recover accidentally-closed exposition popups

### ISSUE THREAD

**Issue by ssokolow** [endless-sky/endless-sky#1697]
I recently introduced my younger (but still University-aged) brother to Endless Sky and, while I was watching him play, I noticed that he'd get so into the flow of rushing through job runs to pay off his initial loan that he'd dismiss exposition popups before he realized what they were.

This is especially troubling since it places the highest chance of receiving "first time on a planet" exposition right at the time when you're most likely to accidentally dismiss it without seeing a way to recover it.

I haven't put much thought into a solution, but the first thing that comes to mind would be, when the user dismisses the dialog, to blank the dialog's contents and then, over the course of less than 300ms (experiment to find the shortest duration that's still clearly recognizable so you don't make the game feel any slower to experienced players), animate the border of the dialog into a clickable square in the bottom-left corner when dismissed. That would provide a clear "click here to undo" cue... especially if paired with some kind of on-hover change in appearance to reinforce that the collapsed square is clickable.

**Comment by jafdy** [comment]
So your saying he clicks through the `diolog` boxes quickly? I'm not aware of any time on a planet where that would cause a problem.

**Comment by ssokolow** [comment]
That depends on what you define as a problem. I consider "Wait. That wasn't an expected popup. What did I miss?" to be a [UX](https://en.wikipedia.org/wiki/User_experience) design flaw.

**Comment by jafdy** [comment]
As far as I know the only non-job diolog boxes on planets are the sort of thing like "the cargo is loaded on to your ship" for story missions.

**Comment by ssokolow** [comment]
I'm aware of those. They're small and centered.

I'm talking about the differently-styled ones that show up in the left third of a 1920x1080 screen after landing, take up roughly the top 2/3rds of said column, and give exposition about the planet. (The kind used for the initial starter mission.)

**Comment by Amazinite** [comment]
@jafdy Conversations, not dialog. 

The only two times I can think of this occurring outside of missions that you accept (and therefore should be paying attention to them anyways) are Remembrance Day and the bombing.

**Comment by jafdy** [comment]
They are called conversation boxes. None of the jobs use the conversation boxes so the fact they are in a different place should alert you to the fact that they are something different.

@Amazinite As I just explained I don't expect a crossover between what you do with diolog and conversation, and as jobs only use diolog boxes this is what I assumed was being talked about.

**Comment by ssokolow** [comment]
@Amazinite That just makes it more important.
@jafdy I observed, first-hand, that crossover does occur.

To paraphrase the common thread flowing through everything a university Human-Computer Interaction course teaches, "You can't patch the human and humans don', so patch the software to work around flaws in the human."

**Comment by jafdy** [comment]
This need some thought...

**Comment by ssokolow** [comment]
That's all I ask.

I suggested the "animated minimize" close behaviour because "make it Undo-able" is generally the best way to solve a problem like this without inconveniencing in the non-error case and "animate it" is a simple, obvious solution for drawing a relationship between two GUI elements.

**Comment by KiLEdEnNis** [comment]
about pop-ups, I remember a type;
go in a system, get an "on enter" message, but unless you touch 0 keys you dismiss it

**Comment by DingusShingleton** [comment]
Ok, a bit of occam's razor here... but what if it was just set so that the only way to dismiss those dialogs was.... to actually click dismiss?

**Comment by Bladewood** [comment]
@DingusShingleton no.

**Comment by MessyMix** [comment]
YES. @KiLEdEnNis WE NEED THIS. I've done this way too much, and especially with the wanderer jump missions I had to load the save. This needs implementation.

**Comment by KiLEdEnNis** [comment]
@DingusShingleton if the only way to dismiss it was to either press dismiss or enter, people could dismiss it w/o having to use the mouse, and if that would apply to "on enter" messages etc. as well that would be a big improvement

**Comment by LocalGod79** [comment]
@DingusShingleton this is a really great solution, except for one factor: this game is completely playable without the mouse...

**Comment by ssokolow** [comment]
@LocalGod79 I think what @DingusShingleton might have meant is to replicate the close behaviour of mature traditional widget toolkits, where not just any keypress will do it, but:

1. There's a keyboard alias for "activate focused widget"
2. The keyboard alias is bound to a real, edge-triggered keypress event (ie. keyboard repeat can't trigger it)
3. The default action button on a dialog is the focused widget by default.

**Comment by MessyMix** [comment]
Sorry, I don't understand the widget bit (not a coder).

But one solution to the point @LocalGod79 made (that the game should be playable without a mouse) would be to use the escape key to dismiss dialog boxes.

**Comment by ssokolow** [comment]
@MessyMix Traditional GUIs are built by clicking together reusable components like buttons, checkboxes, text-entry fields, bits of non-editable text (labels), etc.

"Widget" is the generic term for any of those ready-made "LEGO pieces" that GUIs are built from.

**Comment by endless-sky** [comment]
The trouble is that if there's any keyboard shortcut at all to dismiss dialogs, the player could just jam it to get rid of everything if they don't want to miss it (and may not notice that there's a conversation panel visible in the stack of dialogs). But, providing a keyboard shortcut to close dialogs is mandatory - I don't want players to have to use the mouse.

I don't know of very many programs that let you undo closing a dialog box, though. That doesn't seem like a great solution, either. I think the best solution to this is the "captain's log" idea that's been floating around for a long time - that whenever something important happens, text is added to a log that you can then peruse at your leisure in the player info panel.

**Comment by ssokolow** [comment]
Intuitively cueing the user to understand what does and doesn't go into the log could be tricky to get right but a unified "everything relevant" log is always a good idea in a game.

The problem is that you don't want to clutter up the captain's log with irrelevant fluff, but people may still be curious about what they missed.

That's why I suggested a different mechanism:

* A captain's log serves as a "significant things" filter. You go to it when you have been too busy to play for a week and want to find your place again.
* Browsing flavour text without geographical proximity goes in a codex/-pedia system... but that cuts you off from context.
* The "undo" system I suggested is based around a "I have no idea what that was... but I know where/when it happened. Is that context enough for you?" perspective on the problem.

The only general alternative I can think of to some kind of "undo" mechanism would be some kind of "debug log" view with entries like "Showed 'Welcome to <planet name>'" that are "hyperlinked" to a "Re-display" action... but that'd be even harder to integrate such that users are aware of it when it's needed without it being detrimental to the overall UI design.

As for a more specific alternative, I suppose all use of conversation panels could be audited and then split into two classes: "generates an entry in the captain's log" and discoverable buttons like "About this planet" (in the "landed on the planet" view) which just happen to get auto-clicked on first visit.
